engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/RuntimeExecutionHooks.java

```diff
@@ -21,15 +22,26 @@ public void add(Runnable hook) {
 
   @Override
   public void run() {
+    Set<Runnable> hooksToRun;
     synchronized (hooks) {
-      for (Runnable hook : hooks) {
+      hooksToRun = new LinkedHashSet<>(hooks);
+    }
+
+    Set<Runnable> hooksExecuted = new HashSet<>();
+    try {
+      for (Runnable hook : hooksToRun) {
         try {
           hook.run();
         } catch (Exception e) {
           logger.error("Failed to run execution hook.", e);
+        } finally {
+          hooksExecuted.add(hook);
         }
       }
-      hooks.clear();
+    } finally {
+      synchronized (hooks) {
+        hooks.removeAll(hooksExecuted);
```

Comment:
Unless there is somebody who checks `hooks.isEmpty()` or `hooks.size()` and basis some decision upon such check, then I suggestion to `hooks.clear()` on line 28.

Reply:
I realized that I can simplify the logic :+1: